blockdevice/stats.go

```diff
@@ -220,3 +310,87 @@ func (fs FS) SysBlockDeviceStat(device string) (IOStats, int, error) {
 	}
 	return stat, count, err
 }
+
+func (fs FS) SysBlockDeviceQueueStats(device string) (BlockQueueStats, error) {
+	stat := BlockQueueStats{}
+	// files with uint64 fields
+	for file, p := range map[string]*uint64{
+		"add_random":             &stat.AddRandom,
+		"dax":                    &stat.DAX,
+		"discard_granularity":    &stat.DiscardGranularity,
+		"discard_max_hw_bytes":   &stat.DiscardMaxHWBytes,
+		"discard_max_bytes":      &stat.DiscardMaxBytes,
+		"hw_sector_size":         &stat.HWSectorSize,
+		"io_poll":                &stat.IOPoll,
+		"io_timeout":             &stat.IOTimeout,
+		"iostats":                &stat.IOStats,
+		"logical_block_size":     &stat.LogicalBlockSize,
+		"max_hw_sectors_kb":      &stat.MaxHWSectorsKB,
+		"max_integrity_segments": &stat.MaxIntegritySegments,
+		"max_sectors_kb":         &stat.MaxSectorsKB,
+		"max_segments":           &stat.MaxSegments,
+		"max_segment_size":       &stat.MaxSegmentSize,
+		"minimum_io_size":        &stat.MinimumIOSize,
+		"nomerges":               &stat.NoMerges,
+		"nr_requests":            &stat.NRRequests,
+		"optimal_io_size":        &stat.OptimalIOSize,
+		"physical_block_size":    &stat.PhysicalBlockSize,
+		"read_ahead_kb":          &stat.ReadAHeadKB,
+		"rotational":             &stat.Rotational,
+		"rq_affinity":            &stat.RQAffinity,
+		"write_same_max_bytes":   &stat.WriteSameMaxBytes,
+		"nr_zones":               &stat.NRZones,
+		"chunk_sectors":          &stat.ChunkSectors,
+		"fua":                    &stat.FUA,
+		"max_discard_segments":   &stat.MaxDiscardSegments,
+		"write_zeroes_max_bytes": &stat.WriteZeroesMaxBytes,
+	} {
+		val, err := util.ReadUintFromFile(fs.sys.Path(sysBlockPath, device, sysBlockQueue, file))
+		if err != nil {
+			return stat, err
```

Comment:
If there is an error here we should return a `nil` instead of a partially filled `stat`.

Reply:
For some reason I was thinking this was returning a pointer to a struct so you could return a literal `nil`.  But since this is returning the struct itself I guess it would have been fine to return it as is.

Reply:
Anyway it will be discarded if error is not nil, so i don't think there is much difference what to return in this case.